### DISCUSSION TITLE
COMException when using FFmpegInteropX

### DISCUSSION THREAD

**Discussion by MEK3DK** [ffmpeginteropx/FFmpegInteropX#335] [Q&A] [answered]
I am trying to use FFmpegInteropX in a WinUI 3 packaged project.
I have added the Nuget package FFmpegInteropX 1.1.0-pre38 (which includes .FFmpegUWP (5.1.100) and Microsoft.Graphics.Win2D (1.0.4)).
I have installed FFmpeg (added to path).

But when I run the project I get a COMexception:
![image](https://user-images.githubusercontent.com/31887636/215100050-70671b4c-21ac-4dc7-b1ee-b335c3433c76.png)

I have also tried the samples from this repository, but same problem.
Also tried compilling all of FFmpegInteropX but that ended with build error, related to msys64 and login, thus it should not relate to the COMException problem.

**Comment by lukasf** [comment, marked as answer]
You must add the Microsoft.Windows.CsWinRT nuget package, when using this with .Net Core / WinUI projects.

We will probably release a separate WinUI package at some point, which includes all the references.

**Reply by MEK3DK** [reply]
Already have that package, which is also included in the samples. Sadly it does not help:

![image](https://user-images.githubusercontent.com/31887636/215228925-50cc7d27-46f4-4cb5-bcec-bf6a32059737.png)

**Reply by brabebhin** [reply]
Can you provide a sample repo?

**Reply by MEK3DK** [reply]
Yes I have tried to take everything from the WinUI FFmpegInteropX sample and put it into a single project:

[TryingFFMpegInterop.zip](https://github.com/ffmpeginteropx/FFmpegInteropX/files/10533682/TryingFFMpegInterop.zip)

Since I cannot get past the first lines of FFmpegInteropX related code then I cannot confirm that the project can run/works.
But if you try it out you might not see the COM problem since you likely have the classes registered.

**Reply by MEK3DK** [reply]
Don't know if it matters but I am using Windows 10 Enterprise 22H2 - 19045.2486.

**Reply by brabebhin** [reply]
It is true. it doesn't work for you.
Just as true is that our sample works.

I can't see any obvious differences other than the fact that our sample has a package project that's different from the main application project. You could try replicating that project setup and see if it works for you. Technically the framework should use the package identity in order to gain the priviledge of activating the COM classes. 
There are some version differences between your sample and ours, but matching the versions doesn't seem to do anything.

Unfortunetely, winUI and its coresponding pet project cs/winRT are still very much alpha level quality frameworks, which is why we are not too much in a hurry to support them. cs/winRT in particular does not seem to support some of the features that we need/want, and the development loop is a PITA.

I tested on windows 11, so I don't think windows 10 matters.

**Reply by brabebhin** [reply]
Actually I took the libery to just try that. Using the template (blank app, packaged with windows packaging project - winUI 3 in desktop), the sample code works for me.

Try installing VCRTForwarders.140 nuget package.
It should work for you then. I have no idea why, since you do have a package. 
Oh well.

**Reply by lukasf** [reply]
I also tested this. Indeed it only starts working when adding the forwarders package. Theoretically, it should only be required for unpackaged apps, not for packaged apps. But there seems to be a bug in CsWinRT. As @brabebhin said, the WinUI tech stack is pretty immature right now, despite having a 1.2 version tag.

**Reply by MEK3DK** [reply]
Thank you very much for your help, adding the VCRTForwarders.140 nuget package fixed the problem.
While looking for a solution I had even come across that package, but because it was stated that it is only for unpackaged apps I did not try it. 

Afaik they have changed the WinUI templates such that a specific package project is no longer needed, but I could be wrong.

And yes WinUI development is a complete mess. I follow it closely and in almost all areas they have serious issues.
Which is very sad because having the advantages of WinUI (like this FFmpegInteropX project) is very much needed, since WPF is very outdated. DirectX version, features and so on.
